### PR TITLE
Add clang-tidy integration

### PR DESCRIPTION
**Pull request by ihhub**
Add clang-tidy into our pipeline

### CODE REVIEW

**Comment by oleg-derevenetz** [comment]
Hi @ihhub Really good, I'll take a look at it later today. I have some doubts regarding the need for the workflow to fail, because it is dependent on nature of clang-tidy check set. Sometimes it's better to perform just a PR decoration instead (add a comment to the PR with description of proposed improvements)...

**Comment by ihhub** [comment]
> Hi @ihhub Really good, I'll take a look at it later today. I have some doubts regarding the need for the workflow to fail, because it is dependent on nature of clang-tidy check set. Sometimes it's better to perform just a PR decoration instead (add a comment to the PR with description of proposed improvements)...

I'm okay not to fail. Right now I can't even test these changes without merging it to master :(

**Comment by qarkai** [comment]
AFAIU this action uses default list of checks (`'-*,performance-*,readability-*,bugprone-*,clang-analyzer-*,cppcoreguidelines-*,mpi-*,misc-*'`). There will be a LOT of warnings with this list and it will take much time to run them all. I'd suggest to select required/fixed checks and limit this list.
Also you should probably skip `thirdparty` directory.

**Comment by oleg-derevenetz** [comment]
> There will be a LOT of warnings with this list and it will take much time to run them all. I'd suggest to select required/fixed checks and limit this list.

Yeah. We need to run only carefully selected set of checks, not all of these "matter-of-taste" checks or checks with high rate of false positives. Otherwise we will get a "always-failing-nobody-cares" workflow. If workflow fails, then it should have a good reason for it.

@ihhub

> Right now I can't even test these changes without merging it to master :(

You can try the following: make a subbranch from the `clang-tidy` branch (say, `clang-tidy-test`) and then create a PR to merge `clang-tidy-test` to thev`clang-tidy` (not `master`!). This should work.

**Comment by oleg-derevenetz** [comment]
Hi @qarkai since you are already using clang-tidy for this project, could you suggest a list of checks for this workflow? I suppose clang-analyzer-related checks are not needed, because we already have a separate workflow for clang analyzer, and we don't need checks that will severely break existing codebase, such as the most of `modernize-*` checks.

**Comment by qarkai** [comment]
Hi @oleg-derevenetz I'm not sure if you want me to suggest the specific list of checks or some general recommendations. I'd say that specific list is subjective, would be too long and could be missing some useful checks. So here my opinion on what should be checked.
1) Most of `bugprone-*`, e.g. [bugprone-inaccurate-erase](https://clang.llvm.org/extra/clang-tidy/checks/bugprone-inaccurate-erase.html) would prevent #3830.
2) Some `cppcoreguidelines-*` like [cppcoreguidelines-macro-usage](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-macro-usage.html). There are controversial checks like [cppcoreguidelines-avoid-goto](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-avoid-goto.html) and [cppcoreguidelines-init-variables](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-init-variables.html) and checks aimed for the latest C++ like [cppcoreguidelines-pro-bounds-array-to-pointer-decay](https://clang.llvm.org/extra/clang-tidy/checks/cppcoreguidelines-pro-bounds-array-to-pointer-decay.html) (requires C++20 span).
3) Some `modernize-*` checks. Since this workflow applies to changed code I don't see much harm in using [modernize-loop-convert](https://clang.llvm.org/extra/clang-tidy/checks/modernize-loop-convert.html), [modernize-redundant-void-arg](https://clang.llvm.org/extra/clang-tidy/checks/modernize-redundant-void-arg.html), [modernize-use-bool-literals](https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-bool-literals.html), [modernize-use-using](https://clang.llvm.org/extra/clang-tidy/checks/modernize-use-using.html), etc.
4) Most of `performance-*` checks. Because you can't have too much performance.
5) Some `readability-*` checks, e.g. #3806, #3818. It's matter of taste, also some checks conflicts with code style. e.g. [readability-avoid-const-params-in-decls](https://clang.llvm.org/extra/clang-tidy/checks/readability-avoid-const-params-in-decls.html).

**Review by oleg-derevenetz — APPROVED** [review]
OK, I made this one work, but we still need to settle the set of checks. I suggest that we do this gradually - there is no need to alter this workflow, you just need to change the `.clang-tidy` file in the project root.

Also I still think that there is no need to let this workflow to fail on warnings - in a good way, it is not intended for such use, because the `steps.analyze.outputs.total_comments > 0` condition will be met only on first run, on subsequent runs existing warnings will not be taken into account. I believe, adding a review to the PR is enough in itself.

File: .github/workflows/clang_tidy.yml [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    - uses: actions/checkout@v2
+    - name: Install dependencies
+      run: |
+        sudo apt-get update
+        sudo apt-get install -y libsdl2-dev libsdl2-image-dev libsdl2-mixer-dev libsdl2-ttf-dev
+    - name: Prepare compile_commands.json
+      run: |
+        cmake -B build -DCMAKE_BUILD_TYPE=Release -DUSE_SDL_VERSION=SDL2 -DENABLE_IMAGE=ON -DENABLE_UNICODE=ON -DENABLE_TOOLS=ON -DFHEROES2_STRICT_COMPILATION=ON -DCMAKE_EXPORT_COMPILE_COMMANDS=ON
+        # We need to fixup paths in the compile_commands.json, because clang-tidy-review doesn't work properly with nested directories in the build_dir
+        mv build/compile_commands.json build/compile_commands.json.orig
```
Comment (qarkai) [inline .github/workflows/clang_tidy.yml:24]: Can use `-i/--in-place` option for `sed` instead of renaming.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz and @qarkai are we good to go with this?

**Comment by oleg-derevenetz** [comment]
I think yes, we can add or remove checks in `.clang-tidy` in a process if needed.

**Review by qarkai — APPROVED** [review]

**Comment by qarkai** [comment]
Hi @ihhub I agree with @oleg-derevenetz.

**Comment by ihhub** [comment]
@oleg-derevenetz and @qarkai thank you very much for your corrections and suggestions in this pull request!

**Pull request merged**